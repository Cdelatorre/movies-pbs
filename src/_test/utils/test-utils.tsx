import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { AppStore, RootState, setupStore } from "../../store";
import { BrowserRouter, Router } from "react-router-dom";
import { emptyStoreData } from "./mocks/mock";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = emptyStoreData,

    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
  route?: string
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (route) window.history.pushState({}, "Test routing", route);

    return (
      <BrowserRouter>
        <Provider store={store}>{children}</Provider>;
      </BrowserRouter>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithRouter(ui: React.ReactElement, history?: any) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (history) {
      return (
        <Router location="/" navigator={history}>
          {children}
        </Router>
      );
    } else {
      return <BrowserRouter>{children}</BrowserRouter>;
    }
  }
  return { ...render(ui, { wrapper: Wrapper }) };
}
