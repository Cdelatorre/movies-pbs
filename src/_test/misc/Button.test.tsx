import { fireEvent, screen } from "@testing-library/react";
import { renderWithProviders } from "../utils/test-utils";
import Button from "../../components/misc/button/index";

const actionBtnFn = jest.fn();

describe("Renders Button component correctly", () => {
  test("Renders default Button component correctly", async () => {
    renderWithProviders(
      <Button id="btn-id" action={actionBtnFn}>
        Mock test
      </Button>
    );

    const buttonElement = screen.getByTestId("test-btn-id");

    expect(buttonElement).toBeInTheDocument();
    expect(screen.getByText("Mock test")).toBeInTheDocument();
    expect(buttonElement).toHaveClass("btn btn-primary");
  });

  test("Renders success button type if btnType prop is passed", async () => {
    renderWithProviders(
      <Button id="btn-id" action={actionBtnFn} btnType="success">
        Mock test
      </Button>
    );
    const buttonElement = screen.getByTestId("test-btn-id");

    expect(buttonElement).toHaveClass("btn btn-success");
  });

  test("Fires action function passed as prop if button is clicked", async () => {
    renderWithProviders(
      <Button id="btn-id" action={actionBtnFn}>
        Mock test
      </Button>
    );
    const buttonElement = screen.getByTestId("test-btn-id");

    fireEvent.click(buttonElement);
    expect(actionBtnFn).toHaveBeenCalledTimes(1);
  });
});
