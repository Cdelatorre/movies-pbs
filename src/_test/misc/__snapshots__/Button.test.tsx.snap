// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Button component correctly Renders default Button component correctly 1`] = `
<button
  class="btn btn-primary"
  data-testid="test-btn-id"
  id="btn-id"
>
  Mock test
</button>
`;
